### PR TITLE
Push Notification Action Buttons

### PR DESCRIPTION
**Pull request by azlindsay**
### Description of Changes

- Support for push notification buttons, configured dynamically via the API, added automatically to the alert by SDK.

### Breaking Changes

-   Minimum deployment target bump to iOS 10.0

### Release Checklist

Prepare:

-   [x] Detail any breaking changes. Breaking changes require a new major version number
-   [x] Check `pod lib lint` passes

Bump versions in:

-   [x] `Sources/Kumulos.swift`
-   [x] `KumulosSdkSwift.podspec`
-   [x] `KumulosSdkSwiftExtension.podspec`
-   [x] `Sources/Info-*.plist`
-   [x] `README.md`

Release:

-   [ ] Squash and merge to master
-   [ ] Delete branch once merged
-   [ ] Create tag from master matching chosen version
-   [ ] Run `pod trunk push` to publish to CocoaPods

### CODE REVIEW

File: Sources/Extension/CategoryHelper.swift [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            blocker.signal();
+        }
+        
+        _ = blocker.wait(timeout: DispatchTime.now() + DispatchTimeInterval.seconds(5));
+        
+        return returnedCategories
+    }
+    
+    internal static func getExistingDynamicCategoriesList() -> [String] {
+        let blocker = DispatchSemaphore(value: 1)
```
Comment (cgwyllie) [inline Sources/Extension/CategoryHelper.swift]: Think this semaphore would need to be class level if it is to guard the getter to have only one execution in the critical region at a time.

File: Sources/Kumulos+Push.swift [hunk 2]
```diff
@@ -16,8 +16,15 @@ public class KSPushNotification: NSObject {
     internal(set) open var aps: [AnyHashable:Any]
     internal(set) open var data : [AnyHashable:Any]
     internal(set) open var url: URL?
+    internal(set) open var actionIdentifier: String
```
Comment (cgwyllie) [inline Sources/Kumulos+Push.swift]: How about just making this field optional so handling would be more like:

```
if let actionId = message.actionIdentifier as? String {
  handleButtonAction(message)
} else {
  handleOpen(message)
}
```

File: Sources/Extension/CategoryHelper.swift (hunk 1 above)
Comment (azlindsay) [reply Sources/Extension/CategoryHelper.swift]: Category helper made a singleton and semaphore moved to class level. Not sure if we should introduce a semaphore at the `registerCategory` function so we cant have 2 competing inserts if notifications arrive / are processed multithreaded?

Comment (cgwyllie) [reply Sources/Extension/CategoryHelper.swift]: Hm, good question. I think we would maybe struggle to control that properly as it's the app extension that's woken up, so it runs in its own process.. Not sure if there's already a process active does it handle the next notification or does iOS run a second process?

Comment (cgwyllie) [reply Sources/Extension/CategoryHelper.swift]: Also maybe naming of blocker could be made a little more semantic now it's lifted to class level?

File: Sources/Kumulos+Push.swift [hunk 3]
```diff
@@ -123,25 +108,20 @@ public extension Kumulos {
         Kumulos.trackEvent(eventType: KumulosEvent.MESSAGE_OPENED, properties:params)
     }
 
-    internal func pushHandleOpen(withUserInfo: [AnyHashable: Any]?) {
+    //FIXME? - duplicate this function without the
```
Comment (cgwyllie) [inline Sources/Kumulos+Push.swift]: Relevant or fixed?

**Pull request merged**